### ISSUE TITLE
Distrusting the addon

### ISSUE THREAD

**Issue by Ekleog** [passff/passff#307]
I read the following lines in `passff.py`:
```python
# Make sure we don't allow foreign code execution
if receivedMessage['command'][-4:] != "pass":
    sys.exit(1)
```

So I assume the addon is expected not to be trusted.

So here are a few attack vectors that jumped to my mind:

# Programs that end in `pass`

## Issue

It's possible to execute anything that ends in `pass`, like (unchecked and non-exhaustive) [compass](https://github.com/Compass/compass), [keepass](https://keepass.info/), ksshaskpass, [qtpass](https://qtpass.org/), [rofi-pass](https://github.com/carnager/rofi-pass), spass, [sshpass](https://sourceforge.net/projects/sshpass/), [x11-ssh-askpass](www.jmknoble.net/software/x11-ssh-askpass/)… (just went through all the packages my package manager had that end in `pass`)

It's very unlikely that none of these programs allow arbitrary command injection. And the user is not completely unlikely to have at least some of them installed.

## Solution

Check for a '/' ('\\' on windows) before the trailing `pass`. Or maybe even check that in a controlled environment `${receivedExecutable} --help` actually tells you it's `pass`, which would be even better.

# Unchecked command environment

## Issue

```python
# Set up (modified) command environment
env = dict(os.environ)
if "HOME" not in env:
    env["HOME"] = os.path.expanduser('~')
for key, val in receivedMessage['environment'].items():
    env[key] = val
```

Here is a 0-day (although it's maybe not that important, given I dare hope the add-on is not actually legitimately untrusted, but that still makes one fewer layer of security):
```sh
GIT_EXTERNAL_DIFF='/bin/sh -c "echo Hello World"' pass git diff
```

## Solution

White-list environment variables that are allowed to be set

# Unchecked command arguments

## Issue

```python
cmd = [receivedMessage['command']] + receivedMessage['arguments']
```

Here is another 0-day:
```sh
pass git difftool -x '/bin/sh -c "echo Hello World"' -y
```

## Solution

White-list arguments that are allowed to be set

# Conclusion

Basically, I think there should be *much* more white-listing in this `passff.py`. I think nothing that has not been strictly checked for legitimacy should ever be passed to the subprocess.

As you can see in the above example, things *very* quickly go wrong.

**Comment by tuxor1337** [comment]
Thanks! It's mostly for multi-platform compatibility that we support different pass command with different environment variables and different command line arguments. I would love to restrict everything to a very narrow and comprehensible set of options.

But we have users who want to pipe everything through their custom wrapper scripts so that the path is not really pointing to the actual pass binary, but to the wrapper script. Some users want to pass certain additional command line arguments to pass. And then we have people who want to set very specific environment variables that would cause a ridiculously long whitelist of environment variables.

Nobody could seriously claim to be able to maintain such a long whitelist. But of course you would have to make sure that there are no issues with the entries on that whitelist.

From a security point of view, we should probably restrict to no optional paths, environment variables or command line arguments but force the users to modify the `passff.py` manually according to there needs. I would love to replace the chapter in the PassFF preferences about the "pass script" by a short how-to pointing the user to the file `passff.py` so that she can manually hardcode the modifications she likes. But do you think this is reasonable? Will we face tons of `bug reports` that are caused by uncontrolled modifications of the host app?

**Comment by Ekleog** [comment]
I dare believe that people who would use a custom pass command/arguments/environment would anyway have to know what they are doing at least a bit. As such, having them input things directly in eg. some `pass_command = "pass"`, `arguments = []` and `environment = []` (default values indicated) would likely not break much?

And then it's possible to whitelist exactly the commands the add-on wants to pass to `pass`. Actually, it'd likely be even better to have the add-on not actually passing `pass` commands, but just passing a `I'd like to do [x]` message, and let the `passff.py` script transliterate it to the correct `pass` command -- this way there can be no parsing bug in the `passff.py` script.

**Comment by tuxor1337** [comment]
I moved the host app to an external repository (https://github.com/passff/passff-host) and moved all the `pass` script preferences to a preferences section in the file `passff.py`. What do you think?

**Comment by Ekleog** [comment]
If you're asking me about the code of `passff-host`, this looks great! Just a few points:
 * https://github.com/passff/passff-host/blob/master/src/passff.py#L59 and https://github.com/passff/passff-host/blob/master/src/passff.py#L62 : It would be better to add `"--"` as an argument before the addon-supplied arguments, so that no argument could be injected by the add-on.
 * https://github.com/passff/passff-host/blob/master/src/passff.py#L67 : mostly a matter of defensive programming, but I'd have written this line like `"/" + (key[1:] if key[0] == "/" else key)

Other than that, it looks great to me! So should I expect next release of passff to use this updated script? :D

**Comment by tuxor1337** [comment]
Ah, thanks a lot for the helpful code review! I think that solves this issue.

**Comment by Ekleog** [comment]
Great, thanks!
